src/dfsane.jl

```diff
@@ -0,0 +1,313 @@
+        f(cache.fuₙ, cache.uₙ)
+        f₍ₙₒᵣₘ₎ₙ = norm(cache.fuₙ)^nₑₓₚ
+    end
+
+    if cache.internalnorm(cache.fuₙ) < cache.abstol
+        cache.force_stop = true
+    end
+
+    # Update spectral parameter
+    @. cache.uₙ₋₁ = cache.uₙ - cache.uₙ₋₁
+    @. cache.fuₙ₋₁ = cache.fuₙ - cache.fuₙ₋₁
+
+    α₊ = sum(abs2, cache.uₙ₋₁)
+    @. cache.uₙ₋₁ = cache.uₙ₋₁ * cache.fuₙ₋₁
+    α₋ = sum(cache.uₙ₋₁)
+    cache.σₙ = α₊ / α₋
+
+    # Spectral parameter bounds check
+    if abs(cache.σₙ) > σₘₐₓ || abs(cache.σₙ) < σₘᵢₙ
+        test_norm = sqrt(sum(abs2, cache.fuₙ₋₁))
+        if test_norm > 1
+            cache.σₙ = 1.0
+        elseif testnorm < 1e-5
+            cache.σₙ = 1e5
+        else
+            cache.σₙ = 1.0 / test_norm
+        end
+    end
+
+    # Take step
+    @. cache.uₙ₋₁ = cache.uₙ
+    @. cache.fuₙ₋₁ = cache.fuₙ
+    cache.f₍ₙₒᵣₘ₎ₙ₋₁ = f₍ₙₒᵣₘ₎ₙ
+
+    # Update history
+    cache.ℋ[n % M + 1] = f₍ₙₒᵣₘ₎ₙ
+    cache.stats.nf += 1
+    return nothing
+end
+
+function perform_step!(cache::DFSaneCache{false})
+    @unpack f, alg, f₍ₙₒᵣₘ₎ₙ₋₁, f₍ₙₒᵣₘ₎₀,
+    σₙ, σₘᵢₙ, σₘₐₓ, α₁, γ, τₘᵢₙ, τₘₐₓ, nₑₓₚ, M = cache
+
+    T = eltype(cache.uₙ)
+    n = cache.stats.nsteps
+
+    # Spectral parameter range check
+    σₙ = sign(σₙ) * clamp(abs(σₙ), σₘᵢₙ, σₘₐₓ)
+
+    # Line search direction
+    @. cache.𝒹 = -σₙ * cache.fuₙ₋₁
+
+    η = alg.ηₛ(f₍ₙₒᵣₘ₎₀, n, cache.uₙ₋₁, cache.fuₙ₋₁)
+
+    f̄ = maximum(cache.ℋ)
+    α₊ = α₁
+    α₋ = α₁
+    @. cache.uₙ = cache.uₙ₋₁ + α₊ * cache.𝒹
+
+    cache.fuₙ .= f(cache.uₙ)
+    f₍ₙₒᵣₘ₎ₙ = norm(cache.fuₙ)^nₑₓₚ
+    for _ in 1:(cache.alg.max_inner_iterations)
+        𝒸 = f̄ + η - γ * α₊^2 * f₍ₙₒᵣₘ₎ₙ₋₁
+
+        f₍ₙₒᵣₘ₎ₙ ≤ 𝒸 && break
+
+        α₊ = clamp(α₊^2 * f₍ₙₒᵣₘ₎ₙ₋₁ /
+                   (f₍ₙₒᵣₘ₎ₙ + (T(2) * α₊ - T(1)) * f₍ₙₒᵣₘ₎ₙ₋₁),
+                   τₘᵢₙ * α₊,
+                   τₘₐₓ * α₊)
+        @. cache.uₙ = cache.uₙ₋₁ - α₋ * cache.𝒹
+
+        cache.fuₙ .= f(cache.uₙ)
+        f₍ₙₒᵣₘ₎ₙ = norm(cache.fuₙ)^nₑₓₚ
+
+        f₍ₙₒᵣₘ₎ₙ .≤ 𝒸 && break
+
+        α₋ = clamp(α₋^2 * f₍ₙₒᵣₘ₎ₙ₋₁ / (f₍ₙₒᵣₘ₎ₙ + (T(2) * α₋ - T(1)) * f₍ₙₒᵣₘ₎ₙ₋₁),
+                   τₘᵢₙ * α₋,
+                   τₘₐₓ * α₋)
+
+        @. cache.uₙ = cache.uₙ₋₁ + α₊ * cache.𝒹
+        cache.fuₙ .= f(cache.uₙ)
+        f₍ₙₒᵣₘ₎ₙ = norm(cache.fuₙ)^nₑₓₚ
+    end
+
+    if cache.internalnorm(cache.fuₙ) < cache.abstol
+        cache.force_stop = true
+    end
+
+    # Update spectral parameter
+    @. cache.uₙ₋₁ = cache.uₙ - cache.uₙ₋₁
+    @. cache.fuₙ₋₁ = cache.fuₙ - cache.fuₙ₋₁
+
+    α₊ = sum(abs2, cache.uₙ₋₁)
+    @. cache.uₙ₋₁ = cache.uₙ₋₁ * cache.fuₙ₋₁
+    α₋ = sum(cache.uₙ₋₁)
+    cache.σₙ = α₊ / α₋
+
+    # Spectral parameter bounds check
+    if abs(cache.σₙ) > σₘₐₓ || abs(cache.σₙ) < σₘᵢₙ
+        test_norm = sqrt(sum(abs2, cache.fuₙ₋₁))
+        if test_norm > 1
+            cache.σₙ = 1.0
+        elseif testnorm < 1e-5
+            cache.σₙ = 1e5
+        else
+            cache.σₙ = 1.0 / test_norm
```

Comment:
is this not just a `clamp` op?

Reply:
Yes, you're right haha!

Reply:
Hmm, I don't think so. We are testing on the norm of `cache.fuₙ₋₁` to update `cache.σₙ`. Or maybe you are seeing something I'm not?

Reply:
seems like  a clamp on `1/test_norm` to be between 1 and 1e5

Reply:
AH! Now I get what you're saying. Like: `cache.σₙ = clamp(1.0 / test_norm, 1, 1e5)`
Yes that makes sense.